stdlib/public/core/KeyPath.swift

```diff
@@ -118,12 +121,86 @@ public class AnyKeyPath: Hashable, _AppendKeyPath {
       }
     }
   }
+    
+  /*
+  The following pertains to 32-bit architectures only.
+  We assume everything is a valid pointer to a potential
+  _kvcKeyPathStringPtr except for the first 4KB page which is reserved
+  for the nil pointer. Note that we have to distinguish between a valid
+  keypath offset of 0, and the nil pointer itself.
+  We use maximumOffsetOn32BitArchitecture + 1 for this case.
+    
+  The variable maximumOffsetOn32BitArchitecture is duplicated in the two
+  functions below since having it as a global would make accesses slower,
+  given getOffsetFromStorage() gets called on each KeyPath read. Further,
+  having it as an instance variable in AnyKeyPath would increase the size
+  of AnyKeyPath by 8 bytes.
+  TODO: Find a better method of refactoring this variable if possible.
+  */
+
+  func assignOffsetToStorage(offset: Int) {
+    let maximumOffsetOn32BitArchitecture = 4094
+
+    guard offset >= 0 else {
+      return
+    }
+    // TODO: This just gets the architecture size (32 or 64 bits).
+    // Is there a more efficient way? Something in Builtin maybe?
+    let architectureSize = MemoryLayout<Int>.size
+    if architectureSize == 8 {
+      _kvcKeyPathStringPtr = UnsafePointer<CChar>(bitPattern: -offset - 1)
+    } else {
+      if offset == 0 {
+        _kvcKeyPathStringPtr =
+          UnsafePointer<CChar>(bitPattern: maximumOffsetOn32BitArchitecture + 1)
+      } else if offset < maximumOffsetOn32BitArchitecture {
+        _kvcKeyPathStringPtr = UnsafePointer<CChar>(bitPattern: offset)
+      } else {
+        _kvcKeyPathStringPtr = nil
+      }
+    }
+  }
+
+  // TODO: See if this can be @inlinable since it gets called on each
+  // KeyPath access. It would mean _kvcKeyPathStringPtr would need
+  // to be @usableFromInline. What effect would that have on ABI stability?
+  func getOffsetFromStorage() -> Int? {
+    let maximumOffsetOn32BitArchitecture = 4094
+    guard let storage = _kvcKeyPathStringPtr else {
+      return nil
+    }
+
+    // TODO: Maybe we can get a pointer's raw bits instead of doing
+    // a distance calculation. Note: offsetBase can't be unwrapped
+    // forcefully if its bitPattern is 0x00. Hence the 0x01.
```

Comment:
Yeah, it would be better to get the `bitPattern` of the pointer and do integer arithmetic here, since arithmetic between unrelated pointers is UB.

Reply:
Interesting. I'm looking for a way to get the bit pattern out of a pointer like this (without using `distance(to: )` ) but I'm not finding any methods or computed properties to do that. Is this something that might be added in an extension to UnsafePointers or was this omitted intentionally?

Reply:
You get the bit pattern via `[U]Int(bitPattern: pointer)`.